.github/workflows/docker.yml

```diff
@@ -0,0 +1,95 @@
+name: Build and publish docker images
+on:
+  workflow_call:
+    inputs:
+      arm64:
+        description: If set to true, build an image for arm64 as well as amd64.
+        type: boolean
+        required: false
+        default: false
+      image-name:
+        description: "What to call the docker image. This is the FOO in 'matrixdotorg/FOO:tagname'"
+        type: string
+        required: true
+      dockerfile:
+        description: Path to the Dockerfile, relative to the repository root.
+        type: string
+        required: false
+        default: Dockerfile
+      publish:
+        description: If true, publish to dockerhub.
+        type: boolean
+        required: false
+        default: false
+    secrets:
+      DOCKERHUB_USERNAME:
+        required: true
+      DOCKERHUB_TOKEN:
+        required: true
+
+permissions:
+  contents: read
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up QEMU
+        if: ${{ inputs.arm64 }}
+        id: qemu
+        uses: docker/setup-qemu-action@v1
+        with:
+          platforms: arm64
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Inspect builder
+        run: docker buildx inspect
+          
+      - name: Log in to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      # TODO: consider using https://github.com/docker/metadata-action instead of this
+      # custom magic
+      - name: Calculate docker image tag
+        id: set-tag
+        run: |
+          case "${GITHUB_REF}" in
+              refs/heads/develop)
+                  tag=develop
+                  ;;
+              refs/heads/master|refs/heads/main)
+                  tag=latest
+                  ;;
+              refs/tags/*)
+                  tag=${GITHUB_REF#refs/tags/}
+                  ;;
+              *)
+                  tag=${GITHUB_SHA}
+                  ;;
+          esac
+          echo "::set-output name=tag::$tag"
+
+      - name: Build and push amd64
+        uses: docker/build-push-action@v2
+        with:
+          push: ${{ inputs.publish }}
+          labels: "gitsha1=${{ github.sha }}"
+          tags: "matrixdotorg/${{ inputs.image-name }}:${{ steps.set-tag.outputs.tag }}"
+          file: ${{ inputs.dockerfile }}
+          platforms: linux/amd64
+
+      - name: Build and push arm64
+        if: ${{ inputs.arm64 }}
+        uses: docker/build-push-action@v2
+        with:
+          push: ${{ inputs.publish }}
+          labels: "gitsha1=${{ github.sha }}"
+          tags: "matrixdotorg/${{ inputs.image-name }}:${{ steps.set-tag.outputs.tag }}"
+          file: ${{ inputs.dockerfile }}
+          platforms: linux/arm64
```

Comment:
Have you tried this? I'm not 100% sure, but I think this means the ARM64 image will overwrite the tag for the AMD64 image.

Why not accept the list of platforms as an input? TBH though, I'd be happy to always produce the ARM64 image. It's gaining popularity, it's *so easy* for us to do, but *so annoying* for users of the packages when it's missing.

Reply:
Just caution. I don't want to give the impression that we support running all of our stuff on ARM unless we're all in agreement.

Reply:
> Have you tried this?

Thanks for the pointer to allow a custom name. It works: https://hub.docker.com/r/dmrobertsonelement/sydent

(though it'd be snappier if there were more arm wheels)